### PR TITLE
Add LCAO mw_evaluateValue and mw_evaluateDetRatios

### PR DESCRIPTION
**Pull request by kgasperich**
## Proposed changes
Implement LCAO::mw_evaluateValue using gemm.
Implement LCAO::mw_evaluateDetRatios

Address partly  #4615

## What type(s) of changes does this code introduce?
- New feature

### Does this introduce a breaking change?
- No

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes/No. This PR is up to date with current the current state of 'develop'
- Yes/No. Code added or changed in the PR has been clang-formatted
- Yes/No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes/No. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/QMCWaveFunctions/LCAO/LCAOrbitalSet.cpp [hunk 1]
```diff
@@ -463,6 +465,67 @@ void LCAOrbitalSet::mw_evaluateVGLImplGEMM(const RefVectorWithLeader<SPOSet>& sp
   }
 }
 
+void LCAOrbitalSet::mw_evaluateValue(const RefVectorWithLeader<SPOSet>& spo_list,
+                                     const RefVectorWithLeader<ParticleSet>& P_list,
+                                     int iat,
+                                     const RefVector<ValueVector>& psi_v_list) const
+{
+  OffloadMWVArray phi_v;
```
Comment (ye-luo) [inline src/QMCWaveFunctions/LCAO/LCAOrbitalSet.cpp:473]: Please move this to shared resource.

File: src/QMCWaveFunctions/LCAO/LCAOrbitalSet.cpp [hunk 2]
```diff
@@ -463,6 +465,67 @@ void LCAOrbitalSet::mw_evaluateVGLImplGEMM(const RefVectorWithLeader<SPOSet>& sp
...
+    std::copy_n(phi_v.data_at(iw, 0), output_size, psi_v_list[iw].get().data());
+}
+
+void LCAOrbitalSet::mw_evaluateValueImplGEMM(const RefVectorWithLeader<SPOSet>& spo_list,
+                                             const RefVectorWithLeader<ParticleSet>& P_list,
+                                             int iat,
+                                             OffloadMWVArray& psi_v) const
+{
+  const size_t nw = spo_list.size();
+  OffloadMWVArray phi_v;
```
Comment (ye-luo) [inline src/QMCWaveFunctions/LCAO/LCAOrbitalSet.cpp:490]: Please move this to shared resource.

File: src/QMCWaveFunctions/LCAO/LCAOrbitalSet.cpp [hunk 3]
```diff
@@ -463,6 +465,67 @@ void LCAOrbitalSet::mw_evaluateVGLImplGEMM(const RefVectorWithLeader<SPOSet>& sp
...
+  }
+}
+
+void LCAOrbitalSet::mw_evaluateDetRatios(const RefVectorWithLeader<SPOSet>& spo_list,
+                                         const RefVectorWithLeader<const VirtualParticleSet>& vp_list,
+                                         const RefVector<ValueVector>& psi_list,
+                                         const std::vector<const ValueType*>& invRow_ptr_list,
+                                         std::vector<std::vector<ValueType>>& ratios_list) const
+{
+  size_t nw = spo_list.size();
```
Comment (ye-luo) [inline src/QMCWaveFunctions/LCAO/LCAOrbitalSet.cpp:518]: Add `const`

File: src/QMCWaveFunctions/LCAO/LCAOrbitalSet.cpp [hunk 4]
```diff
@@ -463,6 +465,67 @@ void LCAOrbitalSet::mw_evaluateVGLImplGEMM(const RefVectorWithLeader<SPOSet>& sp
...
+  for (size_t iw = 0; iw < nw; iw++)
+  {
+    for (size_t iat = 0; iat < vp_list[iw].getTotalNum(); iat++)
+    {
+      spo_list[iw].evaluateValue(vp_list[iw], iat, psi_list[iw]);
+      ratios_list[iw][iat] = simd::dot(psi_list[iw].get().data(), invRow_ptr_list[iw], psi_list[iw].get().size());
+    }
+  }
```
Comment (ye-luo) [inline src/QMCWaveFunctions/LCAO/LCAOrbitalSet.cpp:526]: The double loop works for now but eventually it needs to be modified to be offload friendly.

File: src/QMCWaveFunctions/LCAO/SoaLocalizedBasisSet.cpp [hunk 5]
```diff
@@ -216,6 +216,15 @@ void SoaLocalizedBasisSet<COT, ORBT>::evaluateV(const ParticleSet& P, int iat, O
   }
 }
 
+template<class COT, typename ORBT>
+void SoaLocalizedBasisSet<COT, ORBT>::mw_evaluateValue(const RefVectorWithLeader<ParticleSet>& P_list,
+                                                       int iat,
+                                                       OffloadMWVArray& v)
+{
+  for (size_t iw = 0; iw < P_list.size(); iw++)
+    evaluateV(P_list[iw], iat, v.data_at(iw,0));
```
Comment (ye-luo) [inline src/QMCWaveFunctions/LCAO/SoaLocalizedBasisSet.cpp:225]: clang-format?

File: src/QMCWaveFunctions/tests/test_MO.cpp [hunk 6]
```diff
@@ -423,6 +423,11 @@ void test_EtOH_mw(bool transform)
   RefVector<SPOSet::GradVector> dpsi_list   = {dpsi_1, dpsi_2};
   RefVector<SPOSet::ValueVector> d2psi_list = {d2psi_1, d2psi_2};
 
+  size_t nw = psi_list.size();
+  SPOSet::ValueVector psi_v_1(n_mo);
+  SPOSet::ValueVector psi_v_2(n_mo);
+  RefVector<SPOSet::ValueVector> psi_v_list   = {psi_v_1, psi_v_2};
```
Comment (ye-luo) [inline src/QMCWaveFunctions/tests/test_MO.cpp:429]: `RefVector<SPOSet::ValueVector> psi_v_list{psi_v_1, psi_v_2};`

**Review by ye-luo — COMMENTED** [review]
Please retitle this PR to reflect the current implementation. Prefer more frequent PRs.

**Comment by prckent** [comment]
Echoing Ye's comment, we have learned that greater numbers of smaller PRs work best. Once the code presently here is satisfactory, I suggest we merge it and subsequent functionality can come in one or two functions at a time. An additional advantage of this is that the code will be running in the nightlies.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]
LGTM still need non-ANL approval

**Comment by ye-luo** [comment]
Test this please

**Comment by prckent** [comment]
Side question: Is the Identity specialization worth it? Is there a common case where it would be used?

**Comment by prckent** [comment]
Thanks Kevin

**Pull request merged**